### PR TITLE
Support COCO-Wholebody

### PR DESCRIPTION
**Pull request by jin-s13**

### CODE REVIEW

File: mmpose/datasets/datasets/top_down/topdown_coco_wholebody_dataset.py [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+        # sanitize bboxes
+        valid_objs = []
+        for obj in objs:
+            x, y, w, h = obj['bbox']
+            x1 = max(0, x)
+            y1 = max(0, y)
+            x2 = min(width - 1, x1 + max(0, w - 1))
+            y2 = min(height - 1, y1 + max(0, h - 1))
+            if ('area' not in obj
+                    or obj['area'] > 0) and x2 >= x1 and y2 >= y1:
```
Comment (innerlee) [inline mmpose/datasets/datasets/top_down/topdown_coco_wholebody_dataset.py:173]: does zero-width/height make sense, when `x2==x1`?

File: mmpose/datasets/datasets/top_down/topdown_coco_wholebody_dataset.py [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+                'rotation': 0,
+                'joints_3d': joints_3d,
+                'joints_3d_visible': joints_3d_visible,
+                'dataset': self.dataset_name,
+                'bbox_score': 1
+            })
+
+        return rec
+
+    def _coco_keypoint_results_one_category_kernel(self, data_pack):
```
Comment (innerlee) [inline mmpose/datasets/datasets/top_down/topdown_coco_wholebody_dataset.py:207]: looks like the test coverage of this function is low

File: mmpose/datasets/datasets/top_down/topdown_coco_wholebody_dataset.py [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+                img_kpt['center'].tolist(),
+                'scale':
+                img_kpt['scale'].tolist()
+            } for img_kpt, key_point in zip(img_kpts, key_points)]
+
+            cat_results.extend(result)
+
+        return cat_results
+
+    def _do_python_keypoint_eval(self, res_file):
```
Comment (innerlee) [inline mmpose/datasets/datasets/top_down/topdown_coco_wholebody_dataset.py:255]: and this one

File: mmpose/datasets/datasets/top_down/topdown_coco_wholebody_dataset.py [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+
+            result = [{
+                'image_id':
+                img_kpt['image_id'],
+                'category_id':
+                cat_id,
+                'keypoints':
+                key_point[0:self.body_num * 3].tolist(),
+                'foot_kpts':
+                key_point[self.body_num * 3:(self.body_num + self.foot_num) *
```
Comment (innerlee) [inline mmpose/datasets/datasets/top_down/topdown_coco_wholebody_dataset.py:230]: to make this kind of computation short, maybe try this:
```python
cuts = np.cumsum([0, self.body_num, self.foot_num, self.face_num, 21, 21]) * 3
```
and indexing can be written as `key_point[cuts[0]:cuts[1]]`, etc.

File: mmpose/datasets/datasets/top_down/topdown_coco_wholebody_dataset.py (hunk 1 above)
Comment (jin-s13) [reply mmpose/datasets/datasets/top_down/topdown_coco_wholebody_dataset.py:173]: emmm, `x2>x1` seems more reasonable.

File: mmpose/datasets/datasets/top_down/topdown_coco_wholebody_dataset.py (hunk 2 above)
Comment (jin-s13) [reply mmpose/datasets/datasets/top_down/topdown_coco_wholebody_dataset.py:207]: Maybe add the 'unit-test for dataset testing' in a separate PR.

Comment (innerlee) [reply mmpose/datasets/datasets/top_down/topdown_coco_wholebody_dataset.py:207]: fine

**Review by innerlee — APPROVED** [review]

**Comment by innerlee** [comment]
need a pr to update repo's readme, and the change log

**Pull request merged**